### ISSUE TITLE
Improve the entering lobby time

### ISSUE THREAD

**Issue by takahirox** [mozilla/hubs#3565]
**Is your feature request related to a problem? Please describe.**

Entering lobby takes time if there are many objects or avatars in a room especially on low-end device or slow internet because Hubs client fully loads all the necessary assets before entering a lobby.

![image](https://user-images.githubusercontent.com/7637832/102322458-f4ed3380-3f33-11eb-9387-1823a89ed3b1.png)

Long entering time may be a bad experience for users.

**Describe the solution you'd like**

Lazy downloading and background loading would improve the entering lobby time a lot and provide better experience.

A concern is that there may be a chance that "entered a lobby but assets are not ready" might be a bad experience for users. For example, "avatars are not displayed or rendered incorrectly (like lack of textures) but hearing voices" might be stressful. To reduce such stresses, we may need to experiment and check if actually the experience is good.

Probably we need different approaches for different asset types. For example, just lazily downloading background and then starting to render them when they are ready is ok for pictures and videos. But for avatars (and scenes), downloading default light assets before entering a lobby, first rendering them, downloading the real ones background, switching to the real ones when they are ready might be better.

Another concern is this change can make the Hubs client design complex.

**Describe alternatives you've considered**

Optimizing our assets and make them much lighter may be another solution. Pros is we don't need to change the code and can keep the design simple. Cons is no effect to user custom assets.

A mix solution may be letting assets have low resolution data like LOD. First downloading the lower ones and rendering them. Then downloading the real ones background and switching to the real ones when they are ready.

**Additional context**

First of all, we may need to measure the entering time especially on low-end devices and slow internet to know how important solving this problem is.

**Comment by keianhzo** [comment]
A few ideas here:
- LOD info on assets would be good so we can load lowres textureless meshes until highres ones are ready. Also might help in lowend devices
- Show greyed out meshes before textures are loaded. Some translucent grey material or even use a fancy loading shader. Textures are usually the heaviest part of a model so we can show meshes early.
- For avatars maybe load a default avatar until the user's model is ready.

**Comment by takahirox** [comment]
I'm working on LOD. I'm writing a glTF loader plugin https://github.com/takahirox/three-gltf-extensions for glTF LOD extension. It loads all levels simultaneously but the lowest one should be completed first. So we can expect shorter response time.

To use the plugin, we need to replace our glTF loader fork with the official one or bump it into the latest official one. (So I have been working on looking into the replacement lately.) And we need to embed LOD extension to our assets and encourage to use LOD in user custom objects.

To get the best benefit from LOD, .gltf + external files are the best because the embedded ones need to first be entirely downloaded.

**Comment by takahirox** [comment]
>* Show greyed out meshes before textures are loaded. Some translucent grey material or even use a fancy loading shader. Textures are usually the heaviest part of a model so we can show meshes early.
> * For avatars maybe load a default avatar until the user's model is ready.

I like these ideas because it would reduce the stressful I wrote in the issue comment.

> A concern is that there may be a chance that "entered a lobby but assets are not ready" might be a bad experience for users. For example, "avatars are not displayed or rendered incorrectly (like lack of textures) but hearing voices" might be stressful.

**Comment by netpro2k** [comment]
I think one of the main reasons we load everything up front currently is because we had noticed texture uploads causing significant hitching, especially on mobile VR headsets. I think there are better ways we can solve that problem now, including spreading out uploads over multiple frames, and using more compressed textures. This problem was also partially solved when we switched over to using ImageBitmaps.

I would say a decent first pass at this would be only to wait on loading the environment before hiding the loading screen. This will ensure that the navmesh and spawn points are in place. Everything else (video, images, external meshes, etc) is loaded via a `media-loader` so they already naturally has a loading cube that will be shown (we can work on making that better as a second step).

**Comment by johnshaughnessy** [comment]
I am a big, big fan of lazy download. I would like it if chat, mic setup, avatar selection and everything else was available as soon as possible. I'm ok with disallowing room entry and not rendering the scene while the environment is still loading. (We need the environment to know where the spawn points are and I can think of other good reasons like not wanting to have people who are downloading the scene be floating thru walls and such.)

**Comment by keianhzo** [comment]
I agree that lazy loading everything handled by the media-loader is a good start. That would save a bunch of time when scenes are crowded with media and probably improve the user perception in those cases. I guess lazy loading the environment/avatars is more complicated thing that we can discuss on a second pass on this.

**Comment by takahirox** [comment]
> I think there are better ways we can solve that problem now, including spreading out uploads over multiple frames, and using more compressed textures. This problem was also partially solved when we switched over to using ImageBitmaps.

Yes, `ImabeBitmap` and/or Compressed textures should reduce the hitching caused by uploading textures. And I experimented spreading out uploads over multiple frames before and it provided better experience, meaning dropping frame rate to 50fps multiple frames is better than dropping to 20 fps one frame.

https://github.com/takahirox/THREE.TextureUploader

But we should test again especially on low-end devices including mobile VRs to check whether the experience is good.

**Comment by takahirox** [comment]
Regarding the LOD stuffs, ideally the change should be only letting glTF loader register my LOD plugin and adding LOD extensions to our glTF assets. No big code change in Hubs. I hope it would be good start for avatars and environments.
